### PR TITLE
feat!: sync with mainline ABCI

### PR DESCRIPTION
**Pull request by cmwaters**
## Description

This PR modifies the ABCI interface to match that of CometBFT v0.37 and v0.47 of the SDK. There is no utility to extracting out blobs into a separate field within `Data`. 

To get past the problem of the data availabiilty hash and the square size, the last transaction returned by the celestia state machine should always be that square size and the second to last transaction should always be the data hash

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use
  [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: rpc/core/blocks.go [hunk 1]
```diff
@@ -292,17 +292,12 @@ func To32PaddedHexBytes(number uint64) ([]byte, error) {
 // to the hex representation of the square size padded to 32 bytes.
 // For more information, refer to:
 // https://github.com/celestiaorg/quantum-gravity-bridge/blob/master/src/DataRootTuple.sol
-func EncodeDataRootTuple(height uint64, dataRoot [32]byte, squareSize uint64) ([]byte, error) {
+func EncodeDataRootTuple(height uint64, dataRoot []byte) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply rpc/core/blocks.go:295]: there is a PR for that: https://github.com/celestiaorg/quantum-gravity-bridge/pull/145

**Comment by cmwaters** [comment]
Converting this as a draft for now. It seems my original assumption that the `SquareSize` was no longer needed in the consensus node is wrong - we still need it as part of the QGB setup. Will need to think a little more about the design here

File: state/execution.go [hunk 2]
```diff
@@ -131,50 +134,53 @@ func (blockExec *BlockExecutor) CreateProposalBlock(
...
 			panic("block data exceeds max amount of allowed bytes")
 		}
 	}
 
-	newData, err := types.DataFromProto(rawNewData)
-	if err != nil {
-		// todo(evan): see if we can get rid of this panic
-		panic(err)
+	// Celestia passes the data root back as the first transaction returned in
+	// PrepareProposal. Here it is extracted out and set in the header
```
Comment (rootulp) [inline state/execution.go:146]: I'm in favor of this PR! If we adopt it, we may also want to update the line

> Note on the implementation: celestia-app computes the hash in prepare_proposal and returns it to CometBFT via [`blockData.Hash`](https://github.com/celestiaorg/celestia-app/blob/5bbdac2d3f46662a34b2111602b8f964d6e6fba5/app/prepare_proposal.go#L78) so a modification to celestia-core isn't strictly necessary but [comments](https://github.com/celestiaorg/celestia-core/blob/2ec23f804691afc196d0104616e6c880d4c1ca41/types/block.go#L1041-L1042) were added.

in the README here: https://github.com/celestiaorg/celestia-core/blob/9bfa856909861081d2bbc715ee00ed56e9fc43bb/README.md?plain=1#L13

**Review by rootulp — COMMENTED** [review]
No blocking feedback, overall I'm a fan of this!

**Comment by cmwaters** [comment]
Opening this PR back for review again. This is for after mainnet just to be clear. I think we might even want to create a new branch `v0.37.x-celestia` and merge this into that (or use `main`)

**Comment by rootulp** [comment]
> This is for after mainnet just to be clear.

Does anyone have permission to create a new milestone in this repo? It could benefit from a **Post-mainnet** milestone (similar to celestia-app). 

Screenshot 2023-06-15 at 11 58 34 AM

File: abci/types/application.go [hunk 3]
```diff
@@ -98,11 +101,18 @@ func (BaseApplication) ApplySnapshotChunk(req RequestApplySnapshotChunk) Respons
 }
 
 func (BaseApplication) PrepareProposal(req RequestPrepareProposal) ResponsePrepareProposal {
-	return ResponsePrepareProposal{BlockData: req.BlockData}
+	squareSize := len(req.Txs)
```
Comment (rootulp) [inline abci/types/application.go:104]: [question] why is the squareSize set to the number of transactions? I think of these as two distinct values. For example, a block with 100 send transactions could have a square size of 4.

File: abci/types/application.go [hunk 4]
```diff
@@ -98,11 +101,18 @@ func (BaseApplication) ApplySnapshotChunk(req RequestApplySnapshotChunk) Respons
...
+	if squareSize == 0 {
+		squareSize = 1
+	}
+	squareSizeBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(squareSizeBytes, uint64(squareSize))
+	req.Txs = append(req.Txs, tmhash.Sum(nil), squareSizeBytes)
```
Comment (rootulp) [inline abci/types/application.go:110]: [question] is `tmhash.Sum(nil)` a placeholder?

File: abci/types/result.go [hunk 5]
```diff
@@ -43,17 +43,17 @@ func (r ResponseQuery) IsErr() bool {
 
 // IsUnknown returns true if Code is Unknown
 func (r ResponseProcessProposal) IsUnknown() bool {
-	return r.Result == ResponseProcessProposal_UNKNOWN
+	return r.Status == ResponseProcessProposal_UNKNOWN
 }
 
 // IsOK returns true if Code is OK
```
[the remark this answers is not in the document]
Comment (rootulp) [reply abci/types/result.go:49]: [nit] bump @cmwaters b/c the godoc function name should ideally match the actual function name.

File: consensus/mempool_test.go [hunk 6]
```diff
@@ -254,5 +254,13 @@ func (app *CounterApplication) Commit() abci.ResponseCommit {
 
 func (app *CounterApplication) PrepareProposal(
 	req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
-	return abci.ResponsePrepareProposal{BlockData: req.BlockData}
+	dataHash := types.ToTxs(req.Txs).Hash()
+	squareSize := len(req.Txs)
```
Comment (rootulp) [inline consensus/mempool_test.go:258]: same question regarding the relationship between squareSize and req.Txs

File: proto/tendermint/abci/types.proto [hunk 7]
```diff
@@ -127,17 +127,31 @@ message RequestApplySnapshotChunk {
...
-  // block_data is an array of transactions that will be included in a block,
+  // the modified transactions cannot exceed this size.
+  int64 max_tx_bytes = 1;
+  // txs is an array of transactions that will be included in a block,
   // sent to the app for possible modifications.
-  // applications can not exceed the size of the data passed to it.
-  tendermint.types.Data block_data = 1;
-  // If an application decides to populate block_data with extra information, they can not exceed this value.
-  int64 block_data_size = 2;
+  repeated bytes            txs                  = 2;
```
Comment (rootulp) [inline proto/tendermint/abci/types.proto:134]: [question] can the protobuf field `2` be reused? Is it a breaking change if it is reused? Asking b/c prior to this PR field `2` was a in64 for block_data_size

File: proto/tendermint/abci/types.proto [hunk 8]
```diff
@@ -300,18 +314,19 @@ message ResponseApplySnapshotChunk {
   }
 }
 
+// ResponsePrepareProposal reflects the same protobuf message as CometBFT
+// The exception here is that the first transaction is always the hash
```
Comment (rootulp) [inline proto/tendermint/abci/types.proto:318]: [question] Based on code above, I thought the second to last transaction was the Merkle root of the row + col roots and the last transaction was the square size. If so, is this comment inaccurate?

File: state/execution.go [hunk 9]
```diff
@@ -131,50 +135,72 @@ func (blockExec *BlockExecutor) CreateProposalBlock(
 		// purpose for now.
 		panic(err)
 	}
-	rawNewData := preparedProposal.GetBlockData()
+
+	// Celestia passes the data root back as the first transaction and the big endian
```
Comment (rootulp) [inline state/execution.go:139]:
```suggestion
	// Celestia passes the data root back as the second to last transaction and the big endian
```

File: state/execution.go [hunk 10]
```diff
@@ -131,50 +135,72 @@ func (blockExec *BlockExecutor) CreateProposalBlock(
...
+	// encoding of the square size as the second transaction.
```
Comment (rootulp) [inline state/execution.go:140]:
```suggestion
	// encoding of the square size as the last transaction.
```

File: state/execution.go [hunk 11]
```diff
@@ -409,43 +435,62 @@ func execBlockOnProxyApp(
...
+func extendedCommitInfo(c abci.CommitInfo) abci.ExtendedCommitInfo {
+	vs := make([]abci.ExtendedVoteInfo, len(c.Votes))
+	for i := range vs {
+		vs[i] = abci.ExtendedVoteInfo{
+			Validator:       c.Votes[i].Validator,
+			SignedLastBlock: c.Votes[i].SignedLastBlock,
+			/*
+				TODO: Include vote extensions information when implementing vote extensions.
+				VoteExtension:   []byte{},
+			*/
```
Comment (rootulp) [inline state/execution.go:488]: [uber nit] I think it's convention to use `//` for this type of comment

> Go provides C-style /* */ block comments and C++-style // line comments. Line comments are the norm; block comments appear mostly as package comments, but are useful within an expression or to disable large swaths of code.

Source: https://go.dev/doc/effective_go

File: proto/tendermint/abci/types.proto (hunk 8 above)
Comment (cmwaters) [reply proto/tendermint/abci/types.proto:318]: Good catch, I forgot to update the comments when I changed it

File: state/execution.go (hunk 11 above)
Comment (cmwaters) [reply state/execution.go:488]: This is just copied from `cometbft/cometbft` but I don't mind changing it to `//`

File: abci/types/application.go (hunk 3 above)
Comment (rootulp) [reply abci/types/application.go]: [optional] Thanks for clarifying that they are placeholders. If they are expected to be overwritten by celestia-app, then proposal to do something like:

```go
func (BaseApplication) PrepareProposal(req RequestPrepareProposal) ResponsePrepareProposal {
	placeholderDataHash := tmhash.Sum(nil)
	placeholderSquareSizeBytes := make([]byte, 8)
	binary.BigEndian.PutUint64(placeholderSquareSizeBytes, uint64(1))

	// placeholderDataHash and placeholderSquareSizeBytes are expected to be
	// overwritten by celestia-app.
	req.Txs = append(req.Txs, placeholderDataHash, placeholderSquareSizeBytes)
	return ResponsePrepareProposal{Txs: req.Txs}
}
```

File: consensus/mempool_test.go (hunk 6 above)
Comment (rootulp) [reply consensus/mempool_test.go:258]: [nit] can these placeholders look identical to `PrepareProposal` in abci/types/application.go

File: test/e2e/app/app.go [hunk 12]
```diff
@@ -259,12 +261,20 @@ func (app *Application) ApplySnapshotChunk(req abci.RequestApplySnapshotChunk) a
...
-	return abci.ResponsePrepareProposal{BlockData: req.BlockData}
+	dataHash := types.ToTxs(req.Txs).Hash()
+	squareSize := len(req.Txs)
+	if squareSize == 0 {
+		squareSize = 1
+	}
+	squareSizeBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(squareSizeBytes, uint64(squareSize))
+	req.Txs = append(req.Txs, dataHash, squareSizeBytes)
+	return abci.ResponsePrepareProposal{Txs: req.Txs}
```
Comment (rootulp) [inline test/e2e/app/app.go:272]: [nit] can the placeholders here look identically to PrepareProposal in abci/types/application.go

**Review by rootulp — APPROVED** [review]

File: README.md [hunk 13]
```diff
@@ -9,8 +9,8 @@
...
...h` in the block header is determined. In CometBFT, `DataHash` is based on the transactions included in a block. In Celestia, block data (including transactions) are erasure coded into a data square to enable data availability sampling. In order for the header to contain a commitment to this data square, `DataHash` was modified to be the Merkle root of the row and column roots of the erasure coded data square. See [ADR 008](https://github.com/celestiaorg/celestia-core/blob/v0.34.x-celestia/docs/celestia-architecture/adr-008-updating-to-tendermint-v0.35.x.md?plain=1#L20) for the motivation or [celestia-app/pkg/da/data_availability_header.go](https://github.com/celestiaorg/celestia-app/blob/2f89956b22c4c3cfdec19b3b8601095af6f69804/pkg/da/data_availability_header.go) for the implementation. This is computed by the application in `PrepareProposal` and returned to `CometBFT` as the second to last transcation. The last transaction is the big endian encoded uint64 of the square size. This is included in the modified `Data` struct that is gossiped to peers. Similarly `CometBFT` passes the `DataHash` as the second to last tx and the `SquareSize` as the final transaction in `ProcessProposal`.
```
Comment (rootulp) [inline README.md:12]: > This is included in the modified `Data` struct that is gossiped to peers

Does "This" refer to the square size? If so, proposal to replace it with "Square size" for clarity.

[typo] `transcation` => `transaction`

```suggestion
1. Modifications to how `DataHash` in the block header is determined. In CometBFT, `DataHash` is based on the transactions included in a block. In Celestia, block data (including transactions) are erasure coded into a data square to enable data availability sampling. In order for the header to contain a commitment to this data square, `DataHash` was modified to be the Merkle root of the row and column roots of the erasure coded data square. See [ADR 008](https://github.com/celestiaorg/celestia-core/blob/v0.34.x-celestia/docs/celestia-architecture/adr-008-updating-to-tendermint-v0.35.x.md?plain=1#L20) for the motivation or [celestia-app/pkg/da/data_availability_header.go](https://github.com/celestiaorg/celestia-app/blob/2f89956b22c4c3cfdec19b3b8601095af6f69804/pkg/da/data_availability_header.go) for the implementation. This is computed by the application in `PrepareProposal` and returned to `CometBFT` as the second to last transacation. The last transaction is the big endian encoded uint64 of the square size. Square size is included in the modified `Data` struct that is gossiped to peers. Similarly `CometBFT` passes the `DataHash` as the second to last tx and the `SquareSize` as the final transaction in `ProcessProposal`.
```

File: abci/types/application.go [hunk 14]
```diff
@@ -98,11 +101,17 @@ func (BaseApplication) ApplySnapshotChunk(req RequestApplySnapshotChunk) Respons
 }
 
 func (BaseApplication) PrepareProposal(req RequestPrepareProposal) ResponsePrepareProposal {
-	return ResponsePrepareProposal{BlockData: req.BlockData}
+	squareSizeBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(squareSizeBytes, 0)
+
+	// placeholderDataHash and placeholderSquareSizeBytes are expected to be
+	// overwritten by celestia-app.
```
Comment (rootulp) [inline abci/types/application.go:108]: [blocking] it's a bit confusing for this comment to exist b/c it references variables that don't exist.

**Review by rootulp — COMMENTED** [review]
Overall LGTM w/ one blocking comment

**Review by SweeXordious — COMMENTED** [review]
Concerning the PR description:
```
`SquareSize` is also no longer needed...
```
It is still needed for QGB. I guess the description needs to be updated

File: proto/tendermint/abci/types.proto [hunk 15]
```diff
@@ -390,14 +414,20 @@ message VoteInfo {
   bool      signed_last_block = 2;
 }
 
-enum EvidenceType {
+message ExtendedVoteInfo {
+  Validator validator         = 1 [(gogoproto.nullable) = false];
+  bool      signed_last_block = 2;
+  bytes     vote_extension    = 3;  // Reserved for future use
+}
```
Comment (evan-forbes) [inline proto/tendermint/abci/types.proto:421]: is_this_vote_extensions

**Review by evan-forbes — COMMENTED** [review]
do we also need to change the data root to use the second to last hash?
https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/types/block.go#L1025-L1037

otherwise LGTM, should be nice

File: rpc/core/blocks.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (SweeXordious) [reply rpc/core/blocks.go]: Opened this to maybe remove the square size: 
https://github.com/celestiaorg/quantum-gravity-bridge/issues/154

**Comment by cmwaters** [comment]
> It is still needed for QGB. I guess the description needs to be updated

Yes square size is still present. I will update the description

File: state/execution.go (hunk 11 above)
Comment (rootulp) [reply state/execution.go:488]: Oh no worries, we can preserve `/*` style comments if that's what is upstream.

**Review by rootulp — APPROVED** [review]

**Comment by cmwaters** [comment]
> do we also need to change the data root to use the second to last hash?

No because the txs in the data struct don't actually contain the hash. The hash (as well as the square size) remains part of the `Data` struct so the txs in their are simply the transactions to be committed.

The "hack" with the two extra transactions is purely across the ABCI layer

File: rpc/core/blocks.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (SweeXordious) [reply rpc/core/blocks.go]: @cmwaters square size can be removed. Do you want to do it in this PR? or I open a subsequent PR to remove it?

[the remark this answers is not in the document]
Comment (cmwaters) [reply rpc/core/blocks.go]: That's great. Let's do it in a subsequent PR

[the remark this answers is not in the document]
Comment (SweeXordious) [reply rpc/core/blocks.go]: https://github.com/celestiaorg/celestia-core/issues/1040

**Review by evan-forbes — APPROVED** [review]

**Pull request merged**